src/passes/Inlining.cpp

```diff
@@ -158,10 +184,56 @@ struct FunctionInfoScanner
     }
 
     info.size = Measurer::measure(curr->body);
+
+    info.maySkipHeavyWork = maySkipHeavyWork(curr->body);
   }
 
 private:
   NameInfoMap* infos;
+
+  // Check if a function may avoid doing any heavy work at all.
+  bool maySkipHeavyWork(Expression* curr) {
+    Index avoidableWork = 0;
+    Index unavoidableWork = 0;
+
+    auto process = [&](Expression* child) {
+      if (auto* iff = child->dynCast<If>()) {
+        // The condition cannot be avoided.
+        unavoidableWork += CostAnalyzer(iff->condition).cost;
+
+        // Only one arm will be taken.
+        Index ifTrue = CostAnalyzer(iff->ifTrue).cost;
```

Comment:
It seems strange that we are using `Index` for costs. Can we use `uint32_t` or similar instead?

Reply:
Hmm, yeah, this is indeed odd as it isn't proportionate to an index in a wasm module. Let me do that as a followup.